pinot-core/src/main/java/com/linkedin/pinot/core/segment/store/SegmentDirectoryPaths.java

```diff
@@ -21,56 +21,90 @@
+        return new File(indexDir, V3_SUBDIRECTORY_NAME);
       default:
-        throw new UnsupportedOperationException("Segment path for version: " + version +
-            " and segmentIndexDirectory: " + segmentIndexDirectory + " can not be determined ");
+        throw new UnsupportedOperationException(
+            "Unsupported segment version: " + segmentVersion + " while trying to get segment directory from: "
+                + indexDir);
     }
   }
 
-  public static boolean isV3Directory(File path) {
+  public static boolean isV3Directory(@Nonnull File path) {
     return path.toString().endsWith(V3_SUBDIRECTORY_NAME);
   }
 
-  public static @Nullable File findMetadataFile(@Nonnull File indexDir) {
+  @Nullable
+  public static File findMetadataFile(@Nonnull File indexDir, @Nonnull SegmentVersion segmentVersion) {
+    return findFormatFile(indexDir, V1Constants.MetadataKeys.METADATA_FILE_NAME, segmentVersion);
+  }
+
+  @Nullable
+  public static File findMetadataFile(@Nonnull File indexDir) {
     return findFormatFile(indexDir, V1Constants.MetadataKeys.METADATA_FILE_NAME);
   }
 
-  /**
-   * Find the creation metadata file in the segment directory path to
-   * handle different segment formats
-   * @param indexDir index directory to look for file
-   * @return creation metadata file or null if the file is not found
-   */
-  public static @Nullable File findCreationMetaFile(@Nonnull File indexDir) {
+  @Nullable
+  public static File findCreationMetaFile(@Nonnull File indexDir, @Nonnull SegmentVersion segmentVersion) {
+    return findFormatFile(indexDir, V1Constants.SEGMENT_CREATION_META, segmentVersion);
+  }
+
+  @Nullable
+  public static File findCreationMetaFile(@Nonnull File indexDir) {
     return findFormatFile(indexDir, V1Constants.SEGMENT_CREATION_META);
   }
 
-  public static @Nullable File findStarTreeFile(@Nonnull File indexDir) {
+  @Nullable
+  public static File findStarTreeFile(@Nonnull File indexDir, @Nonnull SegmentVersion segmentVersion) {
+    return findFormatFile(indexDir, V1Constants.STAR_TREE_INDEX_FILE, segmentVersion);
+  }
+
+  @Nullable
+  public static File findStarTreeFile(@Nonnull File indexDir) {
     return findFormatFile(indexDir, V1Constants.STAR_TREE_INDEX_FILE);
   }
-  private static @Nullable File findFormatFile(@Nonnull File indexDir, String fileName) {
-    Preconditions.checkNotNull(indexDir);
-    Preconditions.checkArgument(indexDir.exists(), "Path %s does not exist", indexDir);
 
-    if (! indexDir.isDirectory()) {
-      return indexDir;
+  /**
+   * Find a file based on the segment version passed in.
+   * <p>Index directory passed in should be top level segment directory.
+   */
+  @Nullable
+  private static File findFormatFile(@Nonnull File indexDir, @Nonnull String fileName,
+      @Nonnull SegmentVersion segmentVersion) {
+    Preconditions.checkArgument(indexDir.isDirectory(), "Path: %s is not a directory", indexDir);
```

Comment:
indexDir.isDirectory && indexDir.exists() ??  Looking for files in non-existing directory should be error

Non-existing indexDir is an error as against non-existing formatFile for which it returns error.